shared/src/main/scala/com/cibo/evilplot/colors/Color.scala

```diff
@@ -32,50 +32,80 @@ package com.cibo.evilplot.colors
 
 import io.circe.{Decoder, Encoder}
 
+import scala.util.Random
+
 sealed trait Color {
   val repr: String
-  def rgba: (Int,Int,Int,Double)
+  def rgba: (Int, Int, Int, Double)
+  def hsla: HSLA
+  def triadic: (Color, Color)
+  def analogous(offsetDegrees: Double = 14): (Color, Color)
+  def darken(percent: Double): Color
+  def lighten(percent: Double): Color
 }
 
-case object Clear extends Color {
+trait HSLABasedManipulation {
+  def hsla: HSLA
+
+  private def boundHue(hue: Double) = {
+    if (hue < 0) hue + 360 else if (hue >= 360) hue - 360 else hue
+  }
+
+  private def floorCeiling(value: Double)(floor: Double, ceiling: Double) =
+    value.min(ceiling).max(floor)
+
+  def triadic: (HSLA, HSLA) = {
+    val hslaColor = this.hsla
+    (
+      hslaColor.copy(hue = boundHue(hslaColor.hue - 120)),
+      hslaColor.copy(hue = boundHue(hslaColor.hue + 120))
+    )
+  }
+
+  def analogous(offsetDegrees: Double = 14): (HSLA, HSLA) = {
+    val hslaColor = this.hsla
+    (
+      hslaColor.copy(hue = boundHue(hslaColor.hue - offsetDegrees)),
+      hslaColor.copy(hue = boundHue(hslaColor.hue + offsetDegrees))
+    )
+  }
+
+  def darken(percent: Double): HSLA = {
+    val hslaColor = this.hsla
+
+    val newLightness = floorCeiling(hslaColor.lightness - percent)(0, 100)
+    hslaColor.copy(lightness = newLightness)
+  }
+
+  def lighten(percent: Double): HSLA = {
+    val hslaColor = this.hsla
+
+    val newLightness = floorCeiling(hslaColor.lightness + percent)(0, 100)
+    hslaColor.copy(lightness = newLightness)
+  }
+}
+
+case object Clear extends Color with HSLABasedManipulation {
   val repr = "hsla(0, 0%, 0%, 0)"
-  def rgba:(Int,Int,Int,Double) = (0,0,0,0.0)
+  def rgba: (Int, Int, Int, Double) = (0, 0, 0, 0.0)
+  def hsla: HSLA = HSLA(0, 0, 0, 0)
 }
 
-case class HSLA(hue: Int, saturation: Int, lightness: Int, opacity: Double) extends Color {
+case class HSLA(hue: Double, saturation: Double, lightness: Double, opacity: Double)
```

Comment:
also probably a change we want to document

Reply:
Done!